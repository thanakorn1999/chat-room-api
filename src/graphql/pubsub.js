// import { PubSub } from 'apollo-server';
// export const pubsub = new PubSub();

const { PubSub, gql } = require('apollo-server');
const pubsub = new PubSub();

module.exports = pubsub;
