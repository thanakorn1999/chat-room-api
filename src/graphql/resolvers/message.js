const set = require('lodash/set');
const get = require('lodash/get');
const uuid = require('uuid/v1');
const mock = require('../../mock');

module.exports = {
  Query: {
    messages: (parent, { roomName }) => {
      const messages = get(mock, `rooms.${roomName}.messages`, []);
      return messages;
    },
  },
  Mutation: {
    sendMessage: (parent, { roomName, message, name }) => {
      set(mock, `rooms.${roomName}`, {
        messages: [
          ...get(mock, `rooms.${roomName}.messages`, []),
          { id: uuid(), body: message, from: { name: name } },
        ],
      });
      return {
        successful: true,
      };
    },
  },
};
