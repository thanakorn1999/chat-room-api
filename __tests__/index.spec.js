import { createTestClient } from 'apollo-server-testing';
import { ApolloServer, gql } from 'apollo-server';
import { typeDefs } from '../src/graphql/typedefs';
import { resolvers } from '../src/graphql/resolvers';

const GET_MESSAGES = gql`
  query getMessages($roomName: String!) {
    messages(roomName: $roomName) {
      body
    }
  }
`;

const SEND_MESSAGE = gql`
  mutation sendMessage($roomName: String!, $message: String!) {
    sendMessage(roomName: $roomName, message: $message) {
      successful
    }
  }
`;

describe('Message', () => {
  const server = new ApolloServer({
    typeDefs: gql`
      ${typeDefs}
    `,
    resolvers
  });
  const { query, mutate } = createTestClient(server);
  describe('room1', () => {
    it('get messages in room1', async () => {
      const res = await query({
        query: GET_MESSAGES,
        variables: { roomName: 'room1' }
      });
      expect(res.data?.messages).toEqual([
        {
          body: '#1 message body in room1'
        }
      ]);
    });

    it('add message into room1', async () => {
      const res = await mutate({
        query: SEND_MESSAGE,
        variables: { roomName: 'room1', message: '#2 message body in room1' }
      });
      expect(res.data?.sendMessage?.successful).toEqual(true);
    });

    it('messages in room1 should added', async () => {
      const res = await query({
        query: GET_MESSAGES,
        variables: { roomName: 'room1' }
      });
      expect(res.data?.messages).toEqual([
        {
          body: '#1 message body in room1'
        },
        {
          body: '#2 message body in room1'
        }
      ]);
    });
  });

  describe('room2', () => {
    it('get messages in room2', async () => {
      const res = await query({
        query: GET_MESSAGES,
        variables: { roomName: 'room2' }
      });
      expect(res.data?.messages).toEqual([]);
    });

    it('add message into room2', async () => {
      const res = await mutate({
        query: SEND_MESSAGE,
        variables: { roomName: 'room2', message: '#1 message body in room2' }
      });
      expect(res.data?.sendMessage?.successful).toEqual(true);
    });
    it('get messages in room2', async () => {
      const res = await query({
        query: GET_MESSAGES,
        variables: { roomName: 'room2' }
      });
      expect(res.data?.messages).toEqual([
        { body: '#1 message body in room2' }
      ]);
    });
  });
});
