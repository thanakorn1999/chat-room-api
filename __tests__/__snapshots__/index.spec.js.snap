// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetches messages 1`] = `
Object {
  "data": Object {
    "messages": Array [],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
